### PR TITLE
Ferie/feature/ferieind 14

### PR DESCRIPTION
**Pull request by JimmiAgerskov**
Made page for viewing your own vacation reports and possibility to delete/edit a pending report.
Also added approve reports settings (substitutes/personal approver)

### CODE REVIEW

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 1]
```diff
@@ -0,0 +1,103 @@
+﻿module app.vacation {
+    "use strict";
+
+    export abstract class BaseMyVacationReportsController {
+
+        vacationReportsGrid; // Kendo typing is missing some parameters that this controller is using
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: Make your own interface that derives from kendo and add the missing parameters

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            this.personId = $rootScope.CurrentUser.Id;
+            this.loadInitialDate();
+        }
+
+        refreshGrid() {
+            /// <summary>
+            /// Refreshes kendo grid datasource.
+            /// </summary>
+            // Use timeout in case grid is undefined on load.
+            this.$timeout(() => {
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: Kendo has a ready event.
http://docs.telerik.com/kendo-ui/AngularJS/global-events

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+                this.vacationReportsGrid.dataSource.transport.options.read.url = this.getVacationReportsUrl();
+                this.vacationReportsGrid.dataSource.read();
+            });
+        }
+
+        deleteClick = (id) => {
+            this.$modal.open({
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: I recommend moving this logic to a route state instead. Else you're left with a state that you cannot navigate to.
Come see me, I'll show you how solve it.

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+        // Format for datepickers.
+        dateOptions: kendo.ui.DatePickerOptions = {
+            format: "yyyy",
+            start: "decade",
+            depth: "decade"
+        };
+
+        private loadInitialDate() {
+            /// <summary>
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: this appears to be in the wrong location

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+            protected moment,
+            protected ReportStatus) {
+
+            // Set personId. The value on $rootScope is set in resolve in application.js
+            this.personId = $rootScope.CurrentUser.Id;
+            this.loadInitialDate();
+        }
+
+        refreshGrid() {
+            /// <summary>
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: this appears to be in the wrong location

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+    export abstract class BaseMyVacationReportsController {
+
+        vacationReportsGrid; // Kendo typing is missing some parameters that this controller is using
+        vacationReportsOptions: kendo.ui.GridOptions;
+        vacationYear: Date;
+        personId: number;
+
+        constructor(
+            protected $modal,
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: types

File: Presentation.Web/App/Vacation/MyVacationReports/ConfirmDeleteVacationReportController.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    class ConfirmDeleteVacationReportController {
+
+        static $inject: string[] = [
+            "$modalInstance",
+            "itemId",
+            "NotificationService"
+        ];
+
+        constructor(private $modalInstance, public itemId, private NotificationService) {
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/ConfirmDeleteVacationReportController.ts]: types

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+            "$modal",
+            "$rootScope",
+            "VacationReport",
+            "$timeout",
+            "Person",
+            "moment"
+        ];
+
+        constructor(
+            protected $scope,
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: types

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+            protected $modal,
+            protected $rootScope,
+            protected VacationReport,
+            protected $timeout,
+            protected Person,
+            protected moment) {
+
+            super($modal, $rootScope, VacationReport, $timeout, Person, moment, "Accepted");
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: Don't you have an Enum for "Accepted"?

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+
+            this.vacationReportsOptions = {
+                autoBind: false,
+                dataSource: {
+                    type: "odata-v4",
+                    transport: {
+                        read: {
+                            beforeSend(req) {
+                                req.setRequestHeader("Accept", "application/json;odata=fullmetadata");
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: What does that do?

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+                    },
+                    pageSize: 20,
+                    serverPaging: true,
+                    serverAggregates: false,
+                    serverSorting: true,
+                    serverFiltering: true,
+                    sort: { field: "StartTimestamp", dir: "desc" }
+                },
+                pageable: {
+                    messages: {
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: What is the intent here? Is it to give the grid special flavor text or to translate it? It it's only translation then you should just active the da-DK localization

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+                        title: "Ferieafslutning",
+                        template: data => {
+                            const m = this.moment.utc(data.EndTimestamp, 'X');
+                            return `${m._d.getDate()}/${m._d.getMonth() + 1}/${m._d.getFullYear()}`;
+                        }
+                    },
+                    {
+                        template: data => {
+                            if (data.Comment != "") {
+                                return `<button kendo-tooltip k-position="'right'" k-content="'${data.Comment
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: not so great line break

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+                        field: "ApprovedBy.FullName",
+                        title: "Godkendt af"
+                    }
+                ],
+                scrollable: false
+            };
+        }
+
+        getVacationReportsUrl() {
+            return `/odata/VacationReports?status=${this
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: not so great line break

File: Presentation.Web/App/Vacation/MyVacationReports/MyPendingVacationReportsController.ts [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+            "$modal",
+            "$rootScope",
+            "VacationReport",
+            "$timeout",
+            "Person",
+            "moment"
+        ];
+
+        constructor(
+            protected $scope,
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyPendingVacationReportsController.ts]: types

File: Presentation.Web/App/Vacation/MyVacationReports/MyPendingVacationReportsController.ts [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+            protected $modal,
+            protected $rootScope,
+            protected VacationReport,
+            protected $timeout,
+            protected Person,
+            protected moment) {
+            super($modal, $rootScope, VacationReport, $timeout, Person, moment, "Pending");
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyPendingVacationReportsController.ts]: Enum for "Pending"?

File: Presentation.Web/App/Vacation/MyVacationReports/MyRejectedVacationReportsController.ts [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+                                (m._d.getMonth() + 1) +
+                                "/" + // +1 because getMonth is zero indexed.
+                                m._d.getFullYear();
+                        }
+                    },
+                    {
+                        title: "Ferieafslutning",
+                        template: data => {
+                            const m = this.moment.utc(data.EndTimestamp, 'X');
+                            return m._d.getDate() +
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyRejectedVacationReportsController.ts]: Use format instead
http://momentjs.com/docs/#/displaying/

File: Presentation.Web/App/Vacation/MyVacationReports/MyRejectedVacationReportsController.ts [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+                        template: data => {
+                            return data.VacationType === "Regular" ?
+                                "Almindelig ferie" : "6. ferieuge";
+                        }
+                    },
+                    {
+                        field: "CreatedDateTimestamp",
+                        template: data => {
+                            var m = this.moment.utc(data.CreatedDateTimestamp, 'X');
+                            return m._d.getDate() +
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyRejectedVacationReportsController.ts]: Use format instead

File: Presentation.Web/App/Vacation/MyVacationReports/MyRejectedVacationReportsController.ts [hunk 18]
```diff
@@ -0,0 +1,149 @@
...
+                                m._d.getFullYear();
+                        },
+                        title: "Indberettet"
+                    },
+                    {
+                        field: "ClosedDateTimestamp",
+                        title: "Afvist dato",
+                        template: data => {
+                            var m = moment.utc(data.ClosedDateTimestamp, 'X');
+                            var date = m._d.getDate() + "/" +
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyRejectedVacationReportsController.ts]: Use format instead

File: Presentation.Web/App/Vacation/ReportVacation/ReportVacationController.ts [hunk 19]
```diff
@@ -42,10 +45,12 @@
         private maxEndDate: Date;
         private currentUser;
 
-        constructor(private $scope, private Person, private $rootScope, private VacationReport, private NotificationService, private VacationBalanceResource: VacationBalanceResource, private moment, private $modal, private $modalInstance) {
+        constructor(private $scope, private Person, private $rootScope, private VacationReport, private NotificationService, private VacationBalanceResource: VacationBalanceResource, private moment, private $modal, private $modalInstance, private vacationReportId) {
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/ReportVacation/ReportVacationController.ts]: types

File: Presentation.Web/App/Vacation/app.vacation.routes.ts [hunk 20]
```diff
@@ -70,6 +89,24 @@
...
+                        CurrentUser: [
+                            "$rootScope", "Person", ($rootScope, Person) => {
+                                if ($rootScope.CurrentUser == undefined) {
+                                    return Person.GetCurrentUser().$promise.then(res => {
+                                        $rootScope.CurrentUser = res;
+                                    });
+                                } else {
+                                    return $rootScope.CurrentUser;
+                                }
+
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/app.vacation.routes.ts]: remove

File: Presentation.Web/App/Vacation/app.vacation.routes.ts [hunk 21]
```diff
@@ -14,8 +14,7 @@
                     controller: "ReportVacationController",
                     controllerAs: "rvc",
                     resolve: {
-                        adminEditCurrentUser() { return 0; },
-                        ReportId() { return -1; },
+                        vacationReportId () { return 0; },
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/app.vacation.routes.ts]: what is the purpose of this?

File: Presentation.Web/App/Vacation/MyVacationReports/MyVacationReportsView.html [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+        <h1 class="page-header">
+            Mine indberetninger
+        </h1>
+    </div>
+</div>
+
+<div class="row">
+    <div class="col-lg-12">
+        <tabset justified="false">
+            <tab heading="Afventer godkendelse" ng-controller="Vacation.MyPendingVacationReportsController as mvrCtrl" ng-click="mvrCtrl.refreshGrid()">
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/MyVacationReportsView.html]: When using UI-Router you usually don't use ng-controller

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts (hunk 4 above)
Comment (JimmiAgerskov) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: All 3 concrete classes makes use of the loadInitialDate(), care to elaborate?
Same goes for the comment below about refreshGrid()

File: Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts (hunk 11 above)
Comment (JimmiAgerskov) [inline Presentation.Web/App/Vacation/MyVacationReports/MyApprovedVacationReportsController.ts]: The translation is used in every grid, so I guess it is a 'special flavour text'

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts (hunk 4 above)
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: The comment should be above the function :)
Also TS uses JSDoc style comments, I believe.'
http://usejsdoc.org/

File: Presentation.Web/App/Core/Interfaces/KendoGrid.d.ts [hunk 23]
```diff
@@ -0,0 +1,5 @@
+﻿declare namespace  app.core.interfaces {
+    export interface KendoGrid extends kendo.ui.Grid {
+        dataSource: any; // This is the type kendo uses in GridOptions
+    }
+}
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Interfaces/KendoGrid.d.ts]: missing line break

File: Presentation.Web/App/Core/Interfaces/NotificationService.d.ts [hunk 24]
```diff
@@ -0,0 +1,5 @@
+﻿declare namespace  app.core.interfaces {
+    export interface NotificationService {
+        AutoFadeNotification(type, title, message);
+    }
+}
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Interfaces/NotificationService.d.ts]: missing line break

File: Presentation.Web/App/Core/Interfaces/KendoGrid.d.ts [hunk 25]
```diff
@@ -0,0 +1,5 @@
+﻿declare namespace  app.core.interfaces {
+    export interface KendoGrid extends kendo.ui.Grid {
+        dataSource: any; // This is the type kendo uses in GridOptions
```
Comment (snebjorn) [inline Presentation.Web/App/Core/Interfaces/KendoGrid.d.ts]: I believe kendo uses the type kendo.data.DataSource.

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 26]
```diff
@@ -1,36 +1,45 @@
 ﻿module app.vacation {
     "use strict";
 
+    import KendoGrid = core.interfaces.KendoGrid;
+    import Person = core.models.Person;
+
     export abstract class BaseMyVacationReportsController {
 
-        vacationReportsGrid; // Kendo typing is missing some parameters that this controller is using
+        vacationReportsGrid: KendoGrid; // Kendo typing is missing some parameters that this controller is using
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: remove comment

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 27]
```diff
@@ -1,36 +1,45 @@
...
         vacationReportsOptions: kendo.ui.GridOptions;
         vacationYear: Date;
         personId: number;
+        isGridLoaded = false;
 
         constructor(
-            protected $modal,
+            protected $scope,
+            protected $modal: angular.ui.bootstrap.IModalService,
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: isn't it ng.ui...

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts [hunk 28]
```diff
@@ -51,23 +60,7 @@
...
-                windowClass: "app-modal-window-full",
-                resolve: {
-                    vacationReportId: () => {
-                        return id;
-                    }
-                }
-            }).result.then(() => {
-                this.refreshGrid();
-            });
+            this.$state.go(".edit", { vacationReportId: id });
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: no reason to have this in the controller, you can just use ui-sref in the view. It's just a link

File: Presentation.Web/App/Core/Interfaces/KendoGrid.d.ts (hunk 25 above)
Comment (JimmiAgerskov) [inline Presentation.Web/App/Core/Interfaces/KendoGrid.d.ts]: kendo.ui.Grid uses that type, however, kendo.ui.GridOptions uses any. 

Using what you suggests yields the following error:
`Property 'transport' does not exist on type 'DataSource'.`

File: Presentation.Web/App/Vacation/MyVacationReports/View.html [hunk 29]
```diff
@@ -8,9 +8,8 @@
 <div class="container-fluid">
     <div class="row">
         <div class="col-xs-4 col-lg-3 to-front">
-            <input kendo-datepicker culture="'da-DK'" k-options="mvrCtrl.dateOptions" ng-model="mvrCtrol.vacationYearString" k-ng-model="mvrCtrl.vacationYear" ng-change="mvrCtrl.refreshGrid()" />
+            <input kendo-datepicker culture="'da-DK'" k-options="mvrCtrl.dateOptions" ng-model="mvrCtrol.vacationYearString" k-ng-model="mvrCtrl.vacationYear" k-on-change="mvrCtrl.refreshGrid()" />
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/MyVacationReports/View.html]: I'm not sure that `culture="'da-DK'"` works

File: Presentation.Web/App/Vacation/ReportVacation/ReportVacationView.html [hunk 30]
```diff
@@ -7,54 +7,57 @@ <h1 class="page-header">
...
-            <div class="col-md-12">
-                <label class="">
-                    <input type="checkbox" ng-model="rvc.vacationEndsOnFullDay" ng-true-value="false" ng-false-value="true" />
-                    Ferien starter ikke med en hel dag
-                </label>
+            <div class="col-md-6 col-sm-6">
+                <div class="row">
+                    <div class="col-md-12">
+                        <h4>Vælg start dato: <span ng-show="rvc.startDate">{{rvc.startDate | date:"MMM d, yyyy"}}</span></h4>
+                        <kendo-calendar culture="'da-DK'" ng-model="rvc.startDate" k-rebind="endDate" min="rvc.mimimumVacationStartDate"></kendo-calendar>
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/ReportVacation/ReportVacationView.html]: -culture

File: Presentation.Web/App/Vacation/ReportVacation/ReportVacationView.html [hunk 31]
```diff
@@ -7,54 +7,57 @@ <h1 class="page-header">
...
+                            <input type="checkbox" ng-model="rvc.vacationStartsOnFullDay" ng-true-value="false" ng-false-value="true" />
+                            Ferien starter starter midt på dagen:
+                        </label>
+                    </div>
+                </div>
+                <div class="row" ng-hide="rvc.vacationStartsOnFullDay">
+                    <div class="col-md-12">
+                        <h4>Tidspunkt for feriestart</h4>
+                        <p>Udfyldes kun hvis første feriedag ikke er hele dagen</p>
+                        <kendo-time-picker culture="'da-DK'" k-format="'HH:mm'" k-options="rvc.timePickerOptions"
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/ReportVacation/ReportVacationView.html]: -culture

File: Presentation.Web/App/Vacation/ReportVacation/ReportVacationView.html [hunk 32]
```diff
@@ -7,54 +7,57 @@ <h1 class="page-header">
...
+                            <input type="checkbox" ng-model="rvc.vacationEndsOnFullDay" ng-true-value="false" ng-false-value="true" />
+                            Ferien slutter midt på dagen
+                        </label>
+                    </div>
+                </div>
+                <div class="row" ng-hide="rvc.vacationEndsOnFullDay">
+                    <div class="col-md-12">
+                        <h4>Tidspunkt for feriens afslutning</h4>
+                        <p>Udfyldes kun hvis sidste feriedag ikke er hele dagen</p>
+                        <kendo-time-picker culture="'da-DK'" k-format="'HH:mm'" k-options="rvc.timePickerOptions"
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/ReportVacation/ReportVacationView.html]: -culture

File: Presentation.Web/App/Vacation/app.vacation.routes.ts [hunk 33]
```diff
@@ -8,29 +8,35 @@
...
+                .state("vacation.report",
+                {
                     url: "/report",
                     templateUrl: "/App/Vacation/ReportVacation/ReportVacationView.html",
                     controller: "ReportVacationController",
                     controllerAs: "rvc",
                     resolve: {
-                        vacationReportId () { return 0; },
+                        vacationReportId () {
+                            return 0;
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/app.vacation.routes.ts]: this is still strange to me

File: Presentation.Web/App/Vacation/app.vacation.routes.ts [hunk 34]
```diff
@@ -8,29 +8,35 @@
...
+                                        });
                                 } else {
                                     return $rootScope.CurrentUser;
                                 }
-
                             }
                         ],
-                        $modalInstance() { return -1; }
+                        $modalInstance() {
+                            return -1;
```
Comment (snebjorn) [inline Presentation.Web/App/Vacation/app.vacation.routes.ts]: this is also strange

File: Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts (hunk 27 above)
Comment (JimmiAgerskov) [inline Presentation.Web/App/Vacation/MyVacationReports/BaseMyVacationReportsController.ts]: Not with the typings for our version of ui bootstrap.

**Pull request merged**